data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3ObjectWorker.java

```diff
@@ -95,16 +98,22 @@ private void doParseObject(final AcknowledgementSet acknowledgementSet,
         final AtomicInteger saveStateCounter = new AtomicInteger();
         try {
             s3ObjectSize = inputFile.getLength();
+            final Instant lastModifiedTime = inputFile.getLastModified();
+            final Instant now = Instant.now();
+            final Instant originationTime = (lastModifiedTime == null || lastModifiedTime.isAfter(now)) ? now : lastModifiedTime;
 
             codec.parse(inputFile, fileCompressionOption.getDecompressionEngine(), record -> {
                 try {
-                    eventConsumer.accept(record.getData(), s3ObjectReference);
+                    Event event = record.getData();
+                    eventConsumer.accept(event, s3ObjectReference);
+                    event.getMetadata().setExternalOriginationTime(originationTime);
```

Comment:
Rather than modify the metadata here, let's add another `EventConsumer`. This can be a chain of `EventConsumer` classes to modify the event. Otherwise, we will have too many changes here.

Reply:
I am not sure if it makes sense because each consumer needs to be called at different place with different arguments.